block-layer/codecs/aes.md

```diff
@@ -0,0 +1,79 @@
+# Specification: AES CODECS
+
+**Status: Prescriptive - Draft**
+
+This document describes codecs for IPLD blocks (CID + Data) that are encrypted with
+an AES cipher.
+
+The following AES variants are defined in this spec:
+
+| name | multicodec | iv size (in bytes) |
+| --- | --- | --- |
+| aes-gcm | 0x1400 | 12 |
+| aes-cbc | 0x1401 | 16 |
+| aes-ctr | 0x1402 | 12 |
+
+## What this spec is not
+
+This is not a complete system for application privacy. The following issues are
+out of scope for this specification, although they can obviously leverage these codecs:
+
+* Key signaling
+* Access controls
+* Dual-layer encryption w/ replication keys
+
+How you determine what key to apply to an encrypted block will need to be done in the
+application layer. How you decide to encrypt a graph and potentially link the encrypted
+blocks together for replication is done at the application layer. How you manage and access
+keys is done in the application layer.
+
+## Encode/Decode vs Encrypt/Decrypt
+
+The goal of specifying codecs that are used for encryption is to allow the codecs to
+include encryption and decryption programs alongside the codec. However, encrypting and
+decrypting are done by the user and are not done automatically as part of any encode/decode
+operation in the codec.
+
+The encryption program returns a data model value suitable for the block encode program. The
+decode program provides a data model value suitable for the decryption program. And the decryption
+program provides a data model value suitable for parsing into a new block (CID and Bytes). These
+programs are designed to interoperate but it's up to the user to combine them and provide the
+necessary key during encryption and decryption.
+
+## Encrypted Block Format
+
+An encrypted block can be decoded into its initializing vector and the encrypted byte
+payload. Since the initializing vector is a known size for each AES variant the block
+format is simply the iv followed by the byte payload.
+
+```
+| iv | bytes |
+```
+
+This is decoded into IPLD data model of the following schema.
+
+```ipldsch
+type AES struct {
+  iv Bytes
+  bytes Bytes
+} representation map
+```
+
+## Decrypted Block Format
```

Comment:
Maybe it's just me, but this section seems quite out of place.

Is there a codec that turns

```
| uint32(CID byteLength) | CID | Bytes
```

into 

```ipldsch
type DecryptedBlock struct {
  cid Link
  bytes Bytes
} representation map
```

If so, what is it called and where is it used/referenced?

Is an aes-gcm encoded data that properly satisfies `12 byte IV | data bytes` deemed invalid data if when decrypted the data does not match the serialized decrypted block format? If so then this gives us an IPLD format whose correctness is impossible to verify below the application layer which stands outside the norms.

Reply:
We’re in a bit of a grey area as to what “is” and “is not” part of the codec.

Sure, you could say that anything that falls outside the encode/decode function “is not” part of the codec. However:

1) we expect encrypt/decrypt programs to ship with the codec
2) the codec is intended to wrap an IPLD block (CID and Bytes) which is a consistent *concept* across IPLD

However, I **DO NOT** want to overly formalize the representation of the block because, ideally, what is input/output for encrypt/decrypt functions would align with the IPLD library in that language and map to how that library handles blocks, which is **NOT** formalized and there are big differences between implementations.

Reply:
I get what you're saying, although I wonder if we should be disambiguating a bit in this doc by saying what implementers of the codec MUST do (e.g. follow the encoding/decoding rules for the encrypted block format) and what users of the codec SHOULD do (encrypt data of the form presented here, i.e. `| CID | Bytes |`).